client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.js

```diff
@@ -1,129 +1,429 @@
+    // Search for the backup method amongst the currently registered methods
+    const { registeredMethods } = this.state;
+    const backupMethodIndex = registeredMethods.findIndex(
+      candidate => candidate.urlSegment === backupMethod.urlSegment
+    );
+
+    // The backup method isn't even registered
+    if (backupMethodIndex < 0) {
+      return;
     }
 
-    return methods;
+    // Remove the method and set it back into state
+    registeredMethods.splice(backupMethodIndex, 1);
+    this.setState({
+      registeredMethods
+    });
+  }
+
+  /**
+   * Handle a click event to remove the given method. This will prompt the user and call the
+   * relevant endpoint to remove the method
+   *
+   * @param {object} method
+   */
+  async handleClickRemove(method) {
+    const { endpoints: { remove } } = this.props;
+    const { ss: { i18n } } = window;
+
+    // Can't remove if there's no endpoint. UI should prevent this from happening
+    if (!remove) {
+      throw Error('Cannot remove method as no remove endpoint is provided');
+    }
+
+    // Confirm with the user
+    const confirmMessage = i18n._t(
+      'MultiFactorAuthentication.DELETE_CONFIRMATION',
+      fallbacks['MultiFactorAuthentication.DELETE_CONFIRMATION']
+    );
+    const confirmTitle = i18n._t(
+      'MultiFactorAuthentication.CONFIRMATION_TITLE',
+      fallbacks['MultiFactorAuthentication.CONFIRMATION_TITLE']
+    );
+    const buttonLabel = i18n._t(
+      'MultiFactorAuthentication.DELETE_CONFIRMATION_BUTTON',
+      fallbacks['MultiFactorAuthentication.DELETE_CONFIRMATION_BUTTON']
+    );
+
+    if (!await confirm(confirmMessage, { title: confirmTitle, confirmLabel: buttonLabel })) {
+      return;
+    }
+
+    const token = Config.get('SecurityID');
+    const endpoint = `${remove.replace('{urlSegment}', method.urlSegment)}?SecurityID=${token}`;
+
+    fetch(endpoint).then(response => response.json().then(json => {
+      if (response.status === 200) {
+        this.handleRemoveMethod(json.availableMethod);
+
+        if (!json.hasBackupMethod) {
+          this.handleRemoveBackupMethod();
+        }
+
+        return;
+      }
+
+      const message = (json.errors && ` Errors: \n - ${json.errors.join('\n -')}`) || '';
+      throw Error(`Could not delete method. Error code ${response.status}.${message}`);
+    }));
   }
 
+  /**
+   * Handle a request to reset the given method. This will just begin the registration process for
+   * the given method, even if it's not an "availableMethod". The backend endpoint should allow
+   * overwriting the existing registration with a new one.
+   *
+   * @param {object} method
+   */
+  handleResetMethod(method) {
+    const registerableMethod = this.props.allAvailableMethods.find(
+      candidate => candidate.urlSegment === method.urlSegment
+    );
+
+    if (!registerableMethod) {
+      throw Error(`Cannot register the method given: ${method.name} (${method.urlSegment}).`);
```

Comment:
Are these for developers or for the UI? If the latter, should we make them translatable?

Reply:
It's probably a redundant error. The UI should make this impossible - meaning it's an error for devs.